# Fall Prep Notebook 7
## Classifying Pumpkin Seeds I

This notebook has a few exercises that can serve as a primer for what to expect in the `Fall Problem Session 7` notebook. These exercises will touch upon the basic python, `pandas`, `numpy`, `matplotlib` and some supervised learning basic techniques that you may want a refresher on prior to starting `Fall Problem Session 7`.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style("whitegrid")

#### 1. Reading Excel files

##### a.

In the `Data` folder is a file called `Pistachio_16_Features_Dataset.xlsx`. Do a web search to see how you can read in such a file using `pandas`. Store the `DataFrame` in a variable.

##### Sample Solution

In [2]:
df = pd.read_excel("../../Data/Pistachio_16_Features_Dataset.xlsx")

##### b.

Look at a random sample of size $10$ from this data set.

##### Sample Solution

In [3]:
df.sample(10)

AREA  PERIMETER  MAJOR_AXIS  MINOR_AXIS  ECCENTRICITY   EQDIASQ  \
101   89168  1332.2290    504.1752    230.6764        0.8892  336.9454   
1131  75322  1100.2540    429.3439    227.7799        0.8477  309.6820   
723   78792  1216.6970    461.4282    225.2631        0.8727  316.7350   
778   67688  1259.5270    460.9236    200.5296        0.9004  293.5695   
585   44791   993.4760    378.3307    156.2171        0.9108  238.8089   
322   60173  1543.6820    445.3457    183.6638        0.9110  276.7935   
662   58990   953.7090    368.6313    208.9783        0.8238  274.0591   
1715  98268  1211.7010    466.6342    269.8694        0.8158  353.7212   
2058  77159  1294.0580    426.0540    232.4743        0.8380  313.4356   
1431  90638  1165.7321    466.5579    248.9237        0.8458  339.7115   

      SOLIDITY  CONVEX_AREA  EXTENT  ASPECT_RATIO  ROUNDNESS  COMPACTNESS  \
101     0.9708        91854  0.7519        2.1856     0.6313       0.6683   
1131    0.9727        77438  0.7473        1.8849     0.7819       0.7213   
723     0.9625        81864  0.7519        2.0484     0.6688       0.6864   
778     0.9242        73239  0.7102        2.2985     0.5362       0.6369   
585     0.9451        47391  0.7422        2.4218     0.5703       0.6312   
322     0.8804        68345  0.6915        2.4248     0.3173       0.6215   
662     0.9761        60432  0.7687        1.7640     0.8150       0.7435   
1715    0.9917        99089  0.7342        1.7291     0.8411       0.7580   
2058    0.9718        79402  0.7256        1.8327     0.5790       0.7357   
1431    0.9889        91658  0.7516        1.8743     0.8382       0.7281   

      SHAPEFACTOR_1  SHAPEFACTOR_2  SHAPEFACTOR_3  SHAPEFACTOR_4  \
101          0.0057         0.0026         0.4466         0.9762   
1131         0.0057         0.0030         0.5203         0.9806   
723          0.0059         0.0029         0.4712         0.9652   
778          0.0068         0.0030         0.4057         0.9324   
585          0.0084         0.0035         0.3984         0.9649   
322          0.0074         0.0031         0.3863         0.9367   
662          0.0062         0.0035         0.5527         0.9750   
1715         0.0047         0.0027         0.5746         0.9936   
2058         0.0055         0.0030         0.5412         0.9919   
1431         0.0051         0.0027         0.5302         0.9937   

                  Class  
101   Kirmizi_Pistachio  
1131  Kirmizi_Pistachio  
723   Kirmizi_Pistachio  
778   Kirmizi_Pistachio  
585   Kirmizi_Pistachio  
322   Kirmizi_Pistachio  
662   Kirmizi_Pistachio  
1715     Siit_Pistachio  
2058     Siit_Pistachio  
1431     Siit_Pistachio

##### c.

Produce a list of column names from this `DataFrame` using only `pandas`.

##### Sample Solution

In [4]:
df.columns

Index(['AREA', 'PERIMETER', 'MAJOR_AXIS', 'MINOR_AXIS', 'ECCENTRICITY',
       'EQDIASQ', 'SOLIDITY', 'CONVEX_AREA', 'EXTENT', 'ASPECT_RATIO',
       'ROUNDNESS', 'COMPACTNESS', 'SHAPEFACTOR_1', 'SHAPEFACTOR_2',
       'SHAPEFACTOR_3', 'SHAPEFACTOR_4', 'Class'],
      dtype='object')

##### d.

How many of each `Class` value is present in this `DataFrame`? What percentage of the `DataFrame` does each value represent?

##### Sample Solution

In [5]:
df.Class.value_counts()

Kirmizi_Pistachio    1232
Siit_Pistachio        916
Name: Class, dtype: int64

In [6]:
df.Class.value_counts(normalize=True)

Kirmizi_Pistachio    0.573557
Siit_Pistachio       0.426443
Name: Class, dtype: float64

#### 2. Train test splits for classification

##### a.

Write a `for` loop to print out a train test split of the following data `test` 20 times. Set a test set size of $20\%$, do not set a random seed.

In [7]:
test = [0,0,0,0,0,0,0,0,0,0,
           1,1,0,0,0,0,0,1,1,1]

##### Sample Solution

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
for i in range(20):
    print(train_test_split(test,
                              test_size=.2,
                              shuffle=True))

[[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1], [0, 1, 0, 0]]
[[0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0], [0, 0, 0, 0]]
[[0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1], [0, 0, 1, 0]]
[[1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1], [1, 0, 0, 0]]
[[0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0], [0, 0, 0, 0]]
[[0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0], [0, 0, 0, 1]]
[[0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0]]
[[1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0]]
[[0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 1, 0, 0]]
[[1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], [1, 0, 0, 1]]
[[0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0], [0, 1, 0, 1]]
[[1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0], [0, 0, 1, 1]]
[[0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0], [0, 0, 0, 1]]
[[1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0], [0, 0, 0, 0]]
[[1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 1]]
[[0, 0, 0, 0, 0, 0, 0, 1,

##### b.

Did it ever happen that your test set had no `1`s in it?

##### Sample Solution

yes

##### c.

Rerun the loop you ran above, only this time include the argument, `stratify=test`. Are there ever splits where the test set has no `1`s in it this time?

##### Sample Solution

In [10]:
for i in range(20):
    print(train_test_split(test,
                              test_size=.2,
                              shuffle=True,
                              stratify=test))

[[0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0], [0, 1, 0, 0]]
[[0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0], [0, 1, 0, 0]]
[[1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0], [1, 0, 0, 0]]
[[0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0], [0, 0, 0, 1]]
[[0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0], [0, 1, 0, 0]]
[[0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0], [1, 0, 0, 0]]
[[0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0], [1, 0, 0, 0]]
[[0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0], [1, 0, 0, 0]]
[[0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0], [1, 0, 0, 0]]
[[0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0], [0, 0, 1, 0]]
[[0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0], [0, 1, 0, 0]]
[[0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0], [0, 1, 0, 0]]
[[0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0], [0, 1, 0, 0]]
[[0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1], [0, 0, 0, 1]]
[[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1], [0, 1, 0, 0]]
[[1, 0, 0, 0, 0, 0, 0, 0,

##### d.

Read the documentation for `StratifiedKFold` to see how it is different from regular `KFold`. Run the pistachio data set through a stratified k-fold using the `Class` column as the stratification column.

- <a href="https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html">https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html</a>
- <a href="https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html">https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html</a>



##### Sample Solution

In [11]:
from sklearn.model_selection import StratifiedKFold

In [12]:
kfold = StratifiedKFold(shuffle=True,
                           random_state = 2039)

for train_index, test_index in kfold.split(df, df['Class']):
    print(df.iloc[train_index].Class.value_counts(normalize=True))
    print(df.iloc[test_index].Class.value_counts(normalize=True))
    print()
    print()
    print()

Kirmizi_Pistachio    0.573341
Siit_Pistachio       0.426659
Name: Class, dtype: float64
Kirmizi_Pistachio    0.574419
Siit_Pistachio       0.425581
Name: Class, dtype: float64



Kirmizi_Pistachio    0.573341
Siit_Pistachio       0.426659
Name: Class, dtype: float64
Kirmizi_Pistachio    0.574419
Siit_Pistachio       0.425581
Name: Class, dtype: float64



Kirmizi_Pistachio    0.573923
Siit_Pistachio       0.426077
Name: Class, dtype: float64
Kirmizi_Pistachio    0.572093
Siit_Pistachio       0.427907
Name: Class, dtype: float64



Kirmizi_Pistachio    0.573589
Siit_Pistachio       0.426411
Name: Class, dtype: float64
Kirmizi_Pistachio    0.573427
Siit_Pistachio       0.426573
Name: Class, dtype: float64



Kirmizi_Pistachio    0.573589
Siit_Pistachio       0.426411
Name: Class, dtype: float64
Kirmizi_Pistachio    0.573427
Siit_Pistachio       0.426573
Name: Class, dtype: float64





--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2022.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)